File: go/684.redundant-connection.go
/*
 * @lc app=leetcode id=684 lang=golang
 *
 * [684] Redundant Connection
 */

// @lc code=start
func findRedundantConnection(edges [][]int) []int {
	n := len(edges)
	sets := make([]int, n+1)
	for i := 1; i < len(sets); i++ {
		sets[i] = i
	}
	for _, edge := range edges {
		from, to := edge[0], edge[1]
		fromRoot := find(sets, from)
		toRoot := find(sets, to)
		if fromRoot != toRoot {
			union(sets, fromRoot, toRoot)
		} else {
			return edge
		}
	}
	return nil
}

func find(sets []int, i int) int {
	for i != sets[i] {
		i = sets[i]
	}
	return i
}

func union(sets []int, a, b int) {
	sets[b] = a
}
// @lc code=end

File: csharp/684.redundant-connection.cs
/*
 * @lc app=leetcode id=684 lang=csharp
 *
 * [684] Redundant Connection
 */

// @lc code=start

using System.Collections.Generic;

public class Solution 
{
    public int[] FindRedundantConnection(int[][] edges) 
    {
        return this.Kruskal(edges);
    }

    #region Kruskal

    // use Kruskal's MST algorithm to find the edge. 
    // Prim should work too, but less intuitive.
    // edges are weighted by its index.
    public int[] Kruskal(int[][] edges)
    {
        int n = edges.Length;

        // disjoint set for cycle detection
        DisjointSet set = new DisjointSet(n + 1);

        // we need N-1 edge to connect all vertexes
        int edgeCount = 0;
        int nextEdge = 0;
        while (edgeCount < n - 1)
        {
            int from = edges[nextEdge][0];
            int to = edges[nextEdge][1];

            // cycle detection after adding current edge
            int rootFrom = set.Find(from);
            int rootTo = set.Find(to);
            if (rootFrom != rootTo)
            {
                set.Union(from, to);
            }
            else 
            {
                // it is guaranteed to have exact one answer, so we could do this.
                return edges[nextEdge];
            }

            nextEdge++;
        }

        // shouldn't reach here.
        return null;
    }

    // DisjointSet without any optimization
    public class DisjointSet
    {
        private int[] set; 

        public DisjointSet(int size)
        {
            this.set = new int[size];
            for (int i = 0; i < size; i++)
            {
                this.set[i] = i;
            }
        }

        public void Union(int x, int y)
        {
            int rootX = this.Find(x);
            int rootY = this.Find(y);

            this.set[rootX] = rootY;
        }

        public int Find(int x)
        {
            while (this.set[x] != x)
            {
                x = this.set[x];
            }

            return x;
        }
    }

    #endregion Kruskal
}
// @lc code=end

